### DISCUSSION TITLE
Nested progress bars with different columns

### DISCUSSION THREAD

**Discussion by ewels** [Textualize/rich#950] [Q&A] [answered]
I'd love to have two `Progress()` bars working together. I can do this fine by doing `add_task()` twice, but then they have the same layout. If I try using `with Progress() as overall_progress:` twice then the second (nested) one is ignored.

Is there some way around this? Or do I just need to come up with a column structure that works for both progress bars?

Thanks in advance! 🙏🏻 

Phil

**Comment by ewels** [comment]
ps. I did read the [docs](https://rich.readthedocs.io/en/stable/progress.html#customizing) about customisation using the `get_renderables` method but it seems to me that my issue is more about having multiple progress instances rather than customising output - it still takes a single `Process` object. Happy to take another look if that's the right direction though.

**Comment by willmcgugan** [comment]
Hi Phil,

There's currently no straightforward way of having multiple Progress instances running at the same time, but there is a current PR to factor out the Progress renderable so that you could add multiple progress bars (with different layouts) to a [Live](https://rich.readthedocs.io/en/latest/live.html) display. 

Will

**Comment by ewels** [comment]
Ok thanks! And I guess that there is no way to have customisation of the progress bar columns on a task-level instead of Progress instance level?

No problem, just knowing that I'm not missing something is enough for me. I'll figure out a workaround with some custom column variables or something. Thanks for the help!

**Comment by willmcgugan** [comment]
> Ok thanks! And I guess that there is no way to have customisation of the progress bar columns on a task-level instead of Progress instance level?

You could implement `Progress.get_renderables`. The [default](https://github.com/willmcgugan/rich/blob/master/rich/progress.py#L934) creates a single table with all the tasks, but you could render each task individually. 

Something along these lines:

```python
class MyProgress(Progress):
    def get_renderables(self) -> Iterable[RenderableType]:
        for task in self.task:
            task_renderable = render_task(task)
            yield task_renderable
```

When you add a task, you can set arbitrary fields .e.g `progress.add_task("working", foo=1)`. Then in the render function you can refer to those fields with `task.fields['foo']`. You could use that to render some tasks in a different way...

**Comment by ewels** [comment]
Ah, fantastic - thanks! I hadn’t quite twigged that the example in the docs was an internal rich function name and not an arbitrary example name (like `render_task` above.. right?) and so I hadn’t dug through to see how the table was assembled. That together with the way to access fields should be enough to get me started. I’ll let you know how I get on, thank you!

**Comment by ewels** [comment, marked as answer]
Ok great, I have it working! In the end it was super simple - I just overwrite `self.columns` within a custom `Progress.get_renderables` function depending on a task key. Simplified example:

```python
class MyProgress(Progress):
    def get_renderables(self):
        for task in self.tasks:
            if task.fields.get("progress_type") == "mygreenbar":
                self.columns = ("[green]Rich is awesome!", BarColumn())
            if task.fields.get("progress_type") == "mybluebar":
                self.columns = (
                    "[blue]Another bar with a different layout",
                    BarColumn(bar_width=None),
                    "•",
                    DownloadColumn(),
                )
            yield self.make_tasks_table([task])

with MyProgress() as progress:
    taskone = progress.add_task("taskone", progress_type="mygreenbar")
    tasktwo = progress.add_task("tasktwo", progress_type="mybluebar")
```

Note that this change means that you get multiple tables, one for each progress bar, instead of a single table with a row for each status bar. But the whole point is to have varying progress bar columns so I don't think that should matter.

Commit with real code where I implemented this: https://github.com/ewels/nf-core-tools/commit/616d9137a07bb4c5641806096a48caf0f8875d7c

Looks awesome! 🎉 

![Jan-21-2021 22-37-16](https://user-images.githubusercontent.com/465550/105415604-3fe81e00-5c39-11eb-968e-20fd9b858d34.gif)

Many thanks for the help @willmcgugan - you got me on the right track 😀 Hopefully this might be useful for others too.
